diff --git a/src/asl/aslproblem.h b/src/asl/aslproblem.h
index 62c0dd2..488f0bb 100644
--- a/src/asl/aslproblem.h
+++ b/src/asl/aslproblem.h
@@ -287,11 +287,11 @@ class ASLProblem {
 
   // Frees all the arrays that were allocated by modifications to the problem.
   void Free();
-
+public:
   // Write an .nl file.
   void WriteNL(fmt::CStringRef stub, ProblemChanges *pc = 0,
                unsigned flags = 0);
-
+private:
   class Proxy {
    private:
     mutable ASL *asl_;
