diff -u -r scipoptsuite-6.0.2/ug/Makefile /vagrant/scipoptsuite-6.0.2-ddbnb/ug/Makefile
--- scipoptsuite-6.0.2/ug/Makefile	2019-07-08 09:34:03.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/Makefile	2020-09-21 08:22:12.000000000 +0000
@@ -117,6 +117,7 @@
 			ug/paraSolverTerminationState.o \
 			ug/paraSysTimer.o \
 			ug/gzstream.o \
+			ug/ErlPortInterface.o \
 			ug/uggithash.o
 
 
diff -u -r scipoptsuite-6.0.2/ug/src/ug/paraCommMpi.cpp /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraCommMpi.cpp
--- scipoptsuite-6.0.2/ug/src/ug/paraCommMpi.cpp	2019-07-08 09:34:04.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraCommMpi.cpp	2020-10-11 09:21:58.000000000 +0000
@@ -657,13 +657,22 @@
    int count,
    MPI_Datatype datatype,
    int dest,
-   const int tag
+   const int tag,
+   bool block
    )
 {
+    if (block) {
    MPI_CALL (
       MPI_Send( buffer, count, datatype, dest, tag, myComm )
       // MPI_Ssend( buffer, count, datatype, dest, tag, myComm )  // after racing, program hang
    );
+    } else {
+        MPI_Request req;
+   MPI_CALL (
+       MPI_Isend( buffer, count, datatype, dest, tag, myComm, &req)
+      // MPI_Ssend( buffer, count, datatype, dest, tag, myComm )  // after racing, program hang
+   );
+    }
    TAG_TRACE (Send, To, dest, tag);
    return 0;
 }
diff -u -r scipoptsuite-6.0.2/ug/src/ug/paraCommMpi.h /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraCommMpi.h
--- scipoptsuite-6.0.2/ug/src/ug/paraCommMpi.h	2019-07-08 09:34:04.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraCommMpi.h	2020-10-10 20:25:43.000000000 +0000
@@ -241,7 +241,7 @@
 
    /** For Created Datatypes */
    int ubcast( void* buffer, int count, MPI_Datatype datatype, int root );
-   int usend( void* bufer, int count, MPI_Datatype datatype, int dest, int tag );
+   int usend( void* bufer, int count, MPI_Datatype datatype, int dest, int tag, bool block=true );
    int ureceive( void* bufer, int count, MPI_Datatype datatype, int source, int tag );
 
 };
diff -u -r scipoptsuite-6.0.2/ug/src/ug/paraInitiator.h /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraInitiator.h
--- scipoptsuite-6.0.2/ug/src/ug/paraInitiator.h	2019-07-08 09:34:04.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraInitiator.h	2020-09-21 21:15:10.000000000 +0000
@@ -43,6 +43,8 @@
 #include "ugs/ugsParaCommMpi.h"
 #endif
 
+#define UG_WITH_DDBNB
+
 namespace UG
 {
 
@@ -272,6 +274,17 @@
    virtual void writeUgsIncumbentSolution(UGS::UgsParaCommMpi *ugsComm) = 0;
 #endif
 
+#ifdef UG_WITH_DDBNB
+   /** start reading DDBNB incumbent solutions from outside **/
+   virtual void startDDBNBComm() = 0;
+
+   /** write DDBNB incumbent solution **/
+   virtual void writeDDBNBIncumbentSolution() = 0;
+
+   /** check pipes and read DDBNB incumbent solution **/
+   virtual void checkAndReadDDBNBIncumbentSolution() = 0;
+#endif
+
 };
 
 typedef ParaInitiator *ParaInitiatorPtr;
diff -u -r scipoptsuite-6.0.2/ug/src/ug/paraLoadCoordinator.cpp /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraLoadCoordinator.cpp
--- scipoptsuite-6.0.2/ug/src/ug/paraLoadCoordinator.cpp	2019-07-08 09:34:04.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraLoadCoordinator.cpp	2020-10-11 09:25:34.000000000 +0000
@@ -524,6 +524,10 @@
 #ifdef UG_WITH_UGS
          if( commUgs ) checkAndReadIncumbent();
 #endif
+#ifdef UG_WITH_DDBNB
+         paraInitiator->checkAndReadDDBNBIncumbentSolution();
+#endif
+
 
          if( nTerminated == paraSolverPool->getNSolvers() ) break;
       }
@@ -1006,6 +1010,9 @@
          paraInitiator->writeUgsIncumbentSolution(commUgs);
       }
 #endif
+#ifdef UG_WITH_DDBNB
+      paraInitiator->writeDDBNBIncumbentSolution();
+#endif
    }
    else
    {
@@ -3020,6 +3027,9 @@
 #ifdef UG_WITH_UGS
       if( commUgs ) checkAndReadIncumbent();
 #endif
+#ifdef UG_WITH_DDBNB
+      paraInitiator->checkAndReadDDBNBIncumbentSolution();
+#endif
 
       /** completion message may delay */
       if( paraRacingSolverPool && paraRacingSolverPool->getNumActiveSolvers() == 0 )
@@ -3347,6 +3357,9 @@
 #ifdef UG_WITH_UGS
       if( commUgs ) checkAndReadIncumbent();
 #endif
+#ifdef UG_WITH_DDBNB
+      paraInitiator->checkAndReadDDBNBIncumbentSolution();
+#endif
 
       if( paraSolverPool->getNumActiveSolvers() == 0 )
       {
@@ -4817,6 +4830,9 @@
 #ifdef UG_WITH_UGS
       if( commUgs ) checkAndReadIncumbent();
 #endif
+#ifdef UG_WITH_DDBNB
+      paraInitiator->checkAndReadDDBNBIncumbentSolution();
+#endif
 
       /** output tabular solving status */
       if( outputTabularSolvingStatusFlag && (!racingTermination) &&
diff -u -r scipoptsuite-6.0.2/ug/src/ug/paraSolution.h /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraSolution.h
--- scipoptsuite-6.0.2/ug/src/ug/paraSolution.h	2019-07-08 09:34:04.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug/paraSolution.h	2020-10-10 20:26:59.000000000 +0000
@@ -65,7 +65,7 @@
    virtual void bcast(ParaComm *comm, int root) = 0;
 
    /** send solution data to the rank */
-   virtual void send(ParaComm *comm, int destination) = 0;
+   virtual void send(ParaComm *comm, int destination, bool block=true) = 0;
 
    /** receive solution data from the source rank */
    virtual void receive(ParaComm *comm, int source) = 0;
diff -u -r scipoptsuite-6.0.2/ug/src/ug_scip/parascip.cpp /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/parascip.cpp
--- scipoptsuite-6.0.2/ug/src/ug_scip/parascip.cpp	2019-07-08 09:34:05.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/parascip.cpp	2020-09-21 08:22:12.000000000 +0000
@@ -278,6 +278,9 @@
          paraLc = new ParaLoadCoordinator(comm, paraParamSet, paraInitiator, &racingSolversExist, paraTimer, detTimer);
 #endif
       }
+#ifdef UG_WITH_DDBNB
+      paraInitiator->startDDBNBComm();
+#endif
       if( paraInitiator->isWarmStarted() )
       {
 #ifdef UG_WITH_ZLIB
diff -u -r scipoptsuite-6.0.2/ug/src/ug_scip/scipParaInitiator.cpp /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaInitiator.cpp
--- scipoptsuite-6.0.2/ug/src/ug_scip/scipParaInitiator.cpp	2020-10-12 07:21:16.549115813 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaInitiator.cpp	2020-10-10 20:30:05.000000000 +0000
@@ -38,6 +38,12 @@
 #include "scip/debug.h"
 #endif
 
+#ifdef UG_WITH_DDBNB
+#include "ug/ErlPortInterface.h"
+ErlPortInterface g_portInterface;
+class ParaSCIPBestValueAcceptor;
+#endif
+
 // #define UG_SCIP_SOL_FEASIBILITY_CHECK_IN_LC
 
 using namespace UG;
@@ -2358,3 +2364,113 @@
    return;
 }
 #endif
+
+#ifdef UG_WITH_DDBNB
+
+class ParaSCIPBestValueAcceptor : public BestValueAcceptor
+{
+    ParaComm *_paraComm;
+    SCIP *scip;
+public:
+    ParaSCIPBestValueAcceptor(ParaComm *paraComm, SCIP *pScip)
+        : _paraComm(paraComm)
+    {
+        this->scip = pScip;
+    }
+    void acceptNewBestValue(double bestVal, uint16_t seqNumber)
+    {
+        DEF_SCIP_PARA_COMM( scipParaComm, _paraComm );
+        SCIP_SOL *origSol;
+        SCIP_CALL_ABORT( SCIPcreateOrigSol(scip, &origSol, NULL) );
+        SCIP_Bool partial = FALSE;
+        SCIP_Bool error = FALSE;
+        std::ostringstream s;
+        s << "insol-" << seqNumber << ".sol";
+
+        SCIP_CALL_ABORT( SCIPreadSolFile(scip, s.str().c_str(), origSol,
+                FALSE, &partial, &error ) );
+        std::cout << "################## partial = " << partial << ", error = " << error << " ##############" << std::endl;
+      if( (!partial) && (!error) )
+      {
+         SCIP_Bool stored = FALSE;
+         SCIP_CALL_ABORT( SCIPtrySol(scip, origSol, FALSE, TRUE, TRUE, TRUE, FALSE, &stored) );
+
+         std::cout << "################## stored = " << stored << " ##############" << std::endl;
+
+         if( 1 || stored )
+         {
+             SCIP_SOL *sol = origSol;//SCIPgetBestSol(scip);
+            int nVars = SCIPgetNVars(scip);
+            SCIP_VAR **vars = SCIPgetVars(scip);
+            SCIP_Real *vals = new SCIP_Real[nVars];
+            SCIP_CALL_ABORT( SCIPgetSolVals(scip, sol, nVars, vars, vals) );
+            //for (int i = 0 ; i < nVars; ++i) {
+            //    std::cout << i << " "<< vars[i] << " " <<  vals[i] << " " << SCIPvarGetIndex(vars[i]) << " " << SCIPvarGetProbindex(vars[i]) << " " << SCIPvarGetName(vars[i]) << std::endl;
+            //}
+
+            ParaSolution *solution = scipParaComm->createScipParaSolution(
+                           0,
+                           SCIPgetSolTransObj(scip, sol),  // Only this value may be used
+                           nVars,
+                           vars,
+                           vals
+                           );
+            delete [] vals;
+            std::cout << "sol almost sent" << std::endl;
+            solution->send(scipParaComm, 0, false);
+            std::cout << "sol sent" << std::endl;
+         }
+      }
+
+      SCIP_CALL_ABORT( SCIPfreeSol(scip, &origSol) );
+    }
+};
+
+void ScipParaInitiator::startDDBNBComm()
+{
+    acceptor = new ParaSCIPBestValueAcceptor(paraComm, scip);
+    g_portInterface.initialize(true);
+}
+
+void ScipParaInitiator::writeDDBNBIncumbentSolution()
+{
+   /* Write out the solution */
+   seqNumber++;
+   std::ostringstream s;
+   s << "outsol-" << seqNumber << ".sol";
+   FILE *fp = fopen(s.str().c_str(), "w");
+   if( fp == NULL )
+   {
+      fprintf (stderr, "Cannot open solution file to write. File name = %s\n", s.str().c_str());
+      exit(1);
+   }
+
+   SCIP_SOL* sol = SCIPgetBestSol(scip);
+
+   assert(sol);
+   fprintf( fp, "# ");
+
+   SCIP_CALL_ABORT( SCIPprintSol(scip, sol, fp, FALSE) );
+
+   fclose(fp);
+
+   if (!g_portInterface.setBestValue(convertToExternalValue(
+               solution->getObjectiveFuntionValue()), true, seqNumber))
+   {
+       seqNumber--;
+       remove(s.str().c_str());
+   }
+   else{
+     std::cout << "LC: NEW INCUMBENT DDBNB"  << std::endl;
+   }
+   //std::cout << "LC: NEW INCUMBENT ACCEPTED " << sol->getObjectiveFuntionValue() << " " <<
+   //    paraInitiator->convertToExternalValue(sol->getObjectiveFuntionValue()) << std::endl;
+   //paraInitiator->writeDDBNBIncumbentSolution();
+}
+
+void ScipParaInitiator::checkAndReadDDBNBIncumbentSolution()
+{
+    g_portInterface.getBestValue(*acceptor);
+}
+
+#endif
diff -u -r scipoptsuite-6.0.2/ug/src/ug_scip/scipParaInitiator.h /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaInitiator.h
--- scipoptsuite-6.0.2/ug/src/ug_scip/scipParaInitiator.h	2019-07-08 09:34:04.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaInitiator.h	2020-09-21 21:14:42.000000000 +0000
@@ -43,12 +43,15 @@
 #include "ugs/ugsParaCommMpi.h"
 #endif
 
+class ParaSCIPBestValueAcceptor;
+
 namespace ParaSCIP
 {
 
 /** Initiator class */
 class ScipParaInitiator : public UG::ParaInitiator
 {
+public:
    UG::ParaParamSet     *paraParams;
    ScipParaInstance     *instance;
    ScipParaSolution     *solution;
@@ -78,6 +81,10 @@
 #ifdef UG_WITH_UGS
    int                  seqNumber;
 #endif
+#ifdef UG_WITH_DDBNB
+   int                  seqNumber;
+   ParaSCIPBestValueAcceptor *acceptor;
+#endif
 
    bool addRootNodeCuts();
    void outputProblemInfo(int *nNonLinearConsHdlrs);
@@ -96,6 +103,9 @@
 #ifdef UG_WITH_UGS
             , seqNumber(0)
 #endif
+#ifdef UG_WITH_DDBNB
+            , seqNumber(0), acceptor(0)
+#endif
    {
    }
 
@@ -103,6 +113,9 @@
    ~ScipParaInitiator(
          )
    {
+#ifdef UG_WITH_DDBNB
+       //delete acceptor;
+#endif
       if( instance ) delete instance;
       if( solution ) delete solution;
       if( scipDiffParamSetRoot ) delete scipDiffParamSetRoot;
@@ -321,6 +334,17 @@
    void writeUgsIncumbentSolution(UGS::UgsParaCommMpi *ugsComm);
 #endif
 
+#ifdef UG_WITH_DDBNB
+   /** start reading DDBNB incumbent solutions from outside **/
+   virtual void startDDBNBComm();
+
+   /** write DDBNB incumbent solution **/
+   virtual void writeDDBNBIncumbentSolution();
+
+   /** check pipes and read DDBNB incumbent solution **/
+   virtual void checkAndReadDDBNBIncumbentSolution();
+#endif
+
 };
 
 typedef ScipParaInitiator *ScipParaInitiatorPtr;
diff -u -r scipoptsuite-6.0.2/ug/src/ug_scip/scipParaSolution.h /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaSolution.h
--- scipoptsuite-6.0.2/ug/src/ug_scip/scipParaSolution.h	2019-07-08 09:34:05.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaSolution.h	2020-09-21 08:22:12.000000000 +0000
@@ -83,7 +83,7 @@
             // std::cout << "*** index: name: value ***" << std::endl;
             for(int i = 0; i < nVars; i++ )
             {
-               indicesAmongSolvers[i] = SCIPvarGetIndex(vars[i]);
+               indicesAmongSolvers[i] = SCIPvarGetProbindex(vars[i]);
                values[i] = vals[i];
                // std::cout << i << ": " << SCIPvarGetName(vars[indicesAmongSolvers[i]]) << ": " << values[i] << std::endl;
                // std::cout << i << ": " << SCIPvarGetName(vars[i]) << ": " << values[i] << std::endl;
diff -u -r scipoptsuite-6.0.2/ug/src/ug_scip/scipParaSolutionMpi.cpp /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaSolutionMpi.cpp
--- scipoptsuite-6.0.2/ug/src/ug_scip/scipParaSolutionMpi.cpp	2019-07-08 09:34:05.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaSolutionMpi.cpp	2020-10-10 20:21:35.000000000 +0000
@@ -164,7 +164,7 @@
 
 /** send solution data to the rank */
 void
-ScipParaSolutionMpi::send(ParaComm *comm, int destination)
+ScipParaSolutionMpi::send(ParaComm *comm, int destination, bool block)
 {
    DEF_PARA_COMM( commMpi, comm);
    MPI_Datatype preDatatype;
@@ -173,7 +173,7 @@
       MPI_Type_commit( &preDatatype )
    );
    PARA_COMM_CALL(
-      commMpi->usend(&objectiveFunctionValue, 1, preDatatype, destination, TagSolution)
+      commMpi->usend(&objectiveFunctionValue, 1, preDatatype, destination, TagSolution, block)
    );
    MPI_CALL(
       MPI_Type_free( &preDatatype )
@@ -186,7 +186,7 @@
          MPI_Type_commit( &datatype )
       );
       PARA_COMM_CALL(
-            commMpi->usend(indicesAmongSolvers, 1, datatype, destination, TagSolution1)
+            commMpi->usend(indicesAmongSolvers, 1, datatype, destination, TagSolution1, block)
       );
       MPI_CALL(
          MPI_Type_free( &datatype )
diff -u -r scipoptsuite-6.0.2/ug/src/ug_scip/scipParaSolutionMpi.h /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaSolutionMpi.h
--- scipoptsuite-6.0.2/ug/src/ug_scip/scipParaSolutionMpi.h	2019-07-08 09:34:05.000000000 +0000
+++ /vagrant/scipoptsuite-6.0.2-ddbnb/ug/src/ug_scip/scipParaSolutionMpi.h	2020-10-10 20:27:54.000000000 +0000
@@ -82,7 +82,7 @@
    void bcast(UG::ParaComm *comm, int root);
 
    /** send solution data to the rank */
-   void send(UG::ParaComm *comm, int destination);
+   void send(UG::ParaComm *comm, int destination, bool block=true);
 
    /** receive solution data from the source rank */
    void receive(UG::ParaComm *comm, int source);
